// mcp.js/src/abi/index.js
var utils = require('../utils/index');
var EthersAbi = require('./utils/abi-coder').AbiCoder;
const AbiCoder = require('./utils/abi-coder').AbiCoder;

var ethersAbiCoder = new EthersAbi(function (type, value) {
    if (type.match(/^u?int/) && !(utils.judge(value) === 'array') && (!(utils.judge(value) === 'object') || value.constructor.name !== 'BN')) {
        return value.toString();
    }
    return value;
});

// result method
function Result () {
}

/**
 * ABICoder prototype should be used to encode/decode solidity params of any type
 */
class ABICoder {
    /**
     * Encodes the function name to its ABI representation, which are the first 4 bytes of the sha3 of the function name including  types.
     *
     * @method encodeFunctionSignature
     * @param {String|Object} functionName
     * @return {String} encoded function name
     */
    encodeFunctionSignature(functionName) {
        if (utils.judge(functionName) === 'object') {
            functionName = utils._jsonInterfaceMethodToString(functionName);
        }

        return utils.sha3(functionName).slice(0, 10);
    }

    /**
     * Encodes the function name to its ABI representation, which are the first 4 bytes of the sha3 of the function name including  types.
     *
     * @method encodeEventSignature
     * @param {String|Object} functionName
     * @return {String} encoded function name
     */
    encodeEventSignature(functionName) {
        if (utils.judge(functionName) === 'object') {
            functionName = utils._jsonInterfaceMethodToString(functionName);
        }

        return utils.sha3(functionName);
    }

    /**
     * Should be used to encode plain param
     *
     * @method encodeParameter
     * @param {String} type
     * @param {Object} param
     * @return {String} encoded plain param
     */
    encodeParameter(type, param) {
        return this.encodeParameters([type], [param]);
    }

    /**
     * Should be used to encode list of params
     *
     * @method encodeParameters
     * @param {Array} types
     * @param {Array} params
     * @return {String} encoded list of params
     */
    encodeParameters(types, params) {
        return ethersAbiCoder.encode(this.mapTypes(types), params);
    }

    /**
     * Map types if simplified format is used
     *
     * @method mapTypes
     * @param {Array} types
     * @return {Array}
     */
    mapTypes(types) {
        var self = this;
        var mappedTypes = [];
        types.forEach(function (type) {
            if (self.isSimplifiedStructFormat(type)) {
                var structName = Object.keys(type)[0];
                mappedTypes.push(
                    Object.assign(
                        self.mapStructNameAndType(structName),
                        {
                            components: self.mapStructToCoderFormat(type[structName])
                        }
                    )
                );

                return;
            }

            mappedTypes.push(type);
        });

        return mappedTypes;
    }

    /**
     * Check if type is simplified struct format
     *
     * @method isSimplifiedStructFormat
     * @param {string | Object} type
     * @returns {boolean}
     */
    isSimplifiedStructFormat(type) {
        return typeof type === 'object' && typeof type.components === 'undefined' && typeof type.name === 'undefined';
    }

    /**
     * Maps the correct tuple type and name when the simplified format in encode/decodeParameter is used
     *
     * @method mapStructNameAndType
     * @param {string} structName
     * @return {{type: string, name: *}}
     */
    mapStructNameAndType(structName) {
        var type = 'tuple';

        if (structName.indexOf('[]') > -1) {
            type = 'tuple[]';
            structName = structName.slice(0, -2);
        }

        return { type: type, name: structName };
    }

    /**
     * Maps the simplified format in to the expected format of the ABICoder
     *
     * @method mapStructToCoderFormat
     * @param {Object} struct
     * @return {Array}
     */
    mapStructToCoderFormat(struct) {
        var self = this;
        var components = [];
        Object.keys(struct).forEach(function (key) {
            if (typeof struct[key] === 'object') {
                components.push(
                    Object.assign(
                        self.mapStructNameAndType(key),
                        {
                            components: self.mapStructToCoderFormat(struct[key])
                        }
                    )
                );

                return;
            }

            components.push({
                name: key,
                type: struct[key]
            });
        });

        return components;
    }

    /**
     * Encodes a function call from its json interface and parameters.
     *
     * @method encodeFunctionCall
     * @param {Array} jsonInterface
     * @param {Array} params
     * @return {String} The encoded ABI for this function call
     */
    encodeFunctionCall(jsonInterface, params) {
        return this.encodeFunctionSignature(jsonInterface) + this.encodeParameters(jsonInterface.inputs, params).replace('0x', '');
    }

    /**
     * Should be used to decode bytes to plain param
     *
     * @method decodeParameter
     * @param {String} type
     * @param {String} bytes
     * @return {Object} plain param
     */
    decodeParameter(type, bytes) {
        return this.decodeParameters([type], bytes)[0];
    }

    /**
     * Should be used to decode list of params
     *
     * @method decodeParameter
     * @param {Array} outputs
     * @param {String} bytes
     * @return {Array} array of plain params
     */
    decodeParameters(outputs, bytes) {
        if (outputs.length > 0 && (!bytes || bytes === '0x' || bytes === '0X')) {
            throw new Error('Returned values aren\'t valid, did it run Out of Gas?');
        }

        // decode data
        const abiCoder = new AbiCoder();//
        const res = abiCoder.decode(outputs, '0x' + bytes.replace(/0x/i, ''));

        var returnValue = new Result();
        returnValue.__length__ = 0;

        outputs.forEach(function (output, i) {
            if ((utils.judge(output) === 'object') && output.type === 'address') {
                returnValue[output.name || i] = '0x' + res[returnValue.__length__].slice(2);
                // returnValue[output.name || i] = utils.encodeAccount(res[returnValue.__length__].slice(2));
            } else if((utils.judge(output) === 'object') && output.type === 'address[]'){
                var decodedValue = res[returnValue.__length__];
                const accounts=[];
                for(let j=0;j<decodedValue.length;j++){
                    accounts.push('0x' + decodedValue[j].slice(2));
                    // accounts.push(utils.encodeAccount(decodedValue[j].slice(2)) );
                }
                returnValue[output.name || i] = accounts;
            } else {
                let decodedValue = res[returnValue.__length__];
                decodedValue = (decodedValue === '0x') ? null : decodedValue;

                returnValue[i] = decodedValue;
                if ((utils.judge(output) === 'object') && output.name) {
                    returnValue[output.name] = decodedValue;
                }
            }
            returnValue.__length__++;
        });

        return returnValue;
    }

    /**
     * Decodes events non- and indexed parameters.
     *
     * @method decodeLog
     * @param {Object} inputs
     * @param {String} data
     * @param {Array} topics
     * @return {Array} array of plain params
     */
    decodeLog(inputs, data, topics) {
        var _this = this;
        topics = (utils.judge(topics) === 'array') ? topics : [topics];

        data = data || '';

        var notIndexedInputs = [];
        var indexedParams = [];
        var topicCount = 0;

        // TODO check for anonymous logs?
        inputs.forEach(function (input, i) {
            if (input.indexed) {
                indexedParams[i] = (['bool', 'int', 'uint', 'address', 'fixed', 'ufixed'].find(function (staticType) {
                    return input.type.indexOf(staticType) !== -1;
                })) ? _this.decodeParameter(input.type, topics[topicCount]) : topics[topicCount];
                topicCount++;
            } else {
                notIndexedInputs[i] = input;
            }
        });

        var nonIndexedData = data;
        var notIndexedParams = (nonIndexedData) ? this.decodeParameters(notIndexedInputs, nonIndexedData) : [];

        var returnValue = new Result();
        returnValue.__length__ = 0;

        inputs.forEach(function (res, i) {
            returnValue[i] = (res.type === 'string') ? '' : null;

            if (typeof notIndexedParams[i] !== 'undefined') {
                returnValue[i] = notIndexedParams[i];
            }
            if (typeof notIndexedParams[res.name] !== 'undefined') {
                returnValue[i] = notIndexedParams[res.name];
            }
            if (typeof indexedParams[i] !== 'undefined') {
                returnValue[i] = indexedParams[i];
            }

            if (res.name) {
                returnValue[res.name] = returnValue[i];
            }

            returnValue.__length__++;
        });

        return returnValue;
    }
}

module.exports = new ABICoder();
